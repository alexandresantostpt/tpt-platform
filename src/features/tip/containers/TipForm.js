import { connect } from 'react-redux'

import { requestDelAction, requestGetCategoryAction, requestGetSingleAction } from '../actions'
import { requestSaveDataAction } from '@features/library/actions'
import { requestSearchCityGoogleApiAction, requestSearchCityOurApiAction } from '@features/travel/actions'

import { getCategoriesSelector, getObjSelector } from '../selectors'
import { getSuggestionsSelector } from '@features/travel/selectors'

import TipForm from '../pages/TipForm'

const mapStateToProps = state => ({
    categories: getCategoriesSelector(state),
    suggestions: getSuggestionsSelector(state),
    tip: getObjSelector(state)
})

const mapDispatchToProps = {
    requestDelAction,
    requestGetCategoryAction,
    requestGetSingleAction,
    requestSaveDataAction,
    requestSearchCityGoogleApiAction,
    requestSearchCityOurApiAction
}

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(TipForm)
