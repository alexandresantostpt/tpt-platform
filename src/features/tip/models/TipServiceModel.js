import TipLibraryModel from './TipLibraryModel'

class TipServiceModel {
    constructor({ __v, _id, _library, libraries, library, scriptDate }) {
        const newLibraries = libraries ? libraries.map(l => new TipLibraryModel(l)) : []
        this.__v = __v
        this._id = _id
        this._librariesFilter = newLibraries
        this._libraries = newLibraries
        this._library = library ? new TipLibraryModel(library) : null
        this.__library = _library ? new TipLibraryModel(_library) : null
        this._scriptDate = scriptDate
    }

    get v() {
        return this.__v
    }

    get id() {
        return this._id
    }

    get libraries() {
        return this._libraries
    }

    get librariesFilter() {
        return this._librariesFilter
    }

    get library() {
        return this._library
    }

    get libraryObj() {
        return this.__library
    }

    get scriptDate() {
        return this._scriptDate
    }

    filter(payload) {
        this._libraries = this.librariesFilter.filter(library => library.filter(payload))
    }

    getImage() {
        return this.library ? this.library.getImage() : null
    }

    toDTO() {
        const libraryDTO = this.library && this.library.toDTO()
        if (libraryDTO) {
            const {
                city: { _id: id, name, timezone }
            } = this.libraryObj && this.libraryObj.toDTO()
            libraryDTO.city = { id, name, timezone }
        }
        return {
            __v: this.v,
            _id: this.id,
            library: libraryDTO,
            scriptDate: this.scriptDate
        }
    }

    toForm() {
        return {
            __v: this.v,
            _id: this.id,
            library: this.library ? this.library.toForm() : null,
            scriptDate: this.scriptDate || null
        }
    }
}

export default TipServiceModel
