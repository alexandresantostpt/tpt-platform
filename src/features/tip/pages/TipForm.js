import { events } from '@constants/events'

import PropTypes from 'prop-types'
import PubSub from 'pubsub-js'
import React, { useEffect, useRef, useState } from 'react'

import { withRouter } from 'react-router-dom'
import { i18n } from '@i18n'
import { services } from '@/constants/services'
import { getQueryParams } from '@utils/browser'
import { isNotEmpty } from '@utils/functions'
import { colors } from '@helpers/colors'

import Button from '@components/button/Button'
import Cancel from '@components/cancel/Cancel'
import CancelLibrary from '@features/library/components/Cancel'
import FormBuilder from '@components/form/FormBuilder'
import If from '@components/if/If'
import Panel from '@components/panel/Panel'
import FormButtonContainer from '@components/form/container/button/FormButtonContainer'
import WrapperPadding from '@components/wrapper/WrapperPadding'

import Select from '../components/select/Select'
import Tabs from '../components/tabs/Tabs'
import TipServiceModel from '../models/TipServiceModel'
import { tourFormFields, hotelFormFields } from '../fields'

/* eslint-disable complexity */
const TipForm = props => {
    const TABS = {
        hotel: services.HOTEL,
        tip: services.TIP,
        tour: services.TOUR
    }
    const {
        categories,
        formDisabled,
        tip,
        match,
        onCancel,
        redirect,
        requestSearchCityGoogleApiAction,
        requestSearchCityOurApiAction,
        showCancel,
        showTabs,
        suggestions
    } = props
    const {
        params: { id }
    } = match
    const { city, cityId, dateId, form: isForm, scriptDate, scriptId, tipId } = getQueryParams()

    const form = useRef()
    const [image, updateImage] = useState()
    const [category, updateCategory] = useState(null)
    const [subCategory, updateSubCategory] = useState(null)
    const [activeTab, updateActiveTab] = useState('tip')
    const [_formDisabled, _updateFormDisabled] = useState(true)
    const hasImage = !!image || !!tip

    let previousPath = `/travel/${id}/script/tips?scriptId=${scriptId}`
    previousPath += `&scriptDate=${scriptDate}&dateId=${dateId}&tipId=${tipId}&city=${city}&cityId=${cityId}`

    useEffect(() => {
        const { requestGetCategoryAction } = props
        requestGetCategoryAction()
    }, [])

    useEffect(() => {
        _updateFormDisabled(formDisabled)
    }, [formDisabled])

    useEffect(() => {
        _updateFormDisabled(!isForm)
    }, [isForm])

    useEffect(() => {
        const { requestGetSingleAction } = props
        if (tipId && tipId !== 'undefined') {
            requestGetSingleAction({ id: tipId })
        } else {
            requestGetSingleAction({ id: null })
        }
    }, [tipId])

    useEffect(() => {
        if (tip) {
            updateActiveTab(i18n.t(`tips.${tip.type || 'Dica'}`))
        }
    }, [tip])

    const callSaveData = values => {
        const { requestSaveDataAction } = props
        const newValues = { ...values }
        if (category && subCategory) {
            newValues.library.category = category.describe
            newValues.library.subCategory = subCategory.describe
        }
        requestSaveDataAction({
            data: {
                dateId,
                scriptId,
                service: new TipServiceModel({ ...newValues, scriptDate }).toDTO(),
                type: TABS[activeTab]
            },
            image,
            redirectTo: redirect ? previousPath : null
        })
        reset()
    }

    const del = () => {
        const { requestDelAction } = props
        requestDelAction({ id: tipId, redirectTo: previousPath })
        reset()
    }

    const onSaveClick = () => form.current.form.submit()

    const reset = () => {
        PubSub.publish(events.FORM_RESET)
        updateImage(null)
    }

    const handleSelectCategory = (categorySelected, subCategorySelected) => {
        updateCategory(categorySelected)
        updateSubCategory(subCategorySelected)
    }

    const renderCancel = () => {
        if (showCancel) {
            if (activeTab === 'tip') {
                if (isNotEmpty(category)) {
                    // eslint-disable-next-line
                    if (onCancel) {
                        return <CancelLibrary onClick={onCancel} />
                    }
                    return <Cancel center link={previousPath} />
                }
            } else {
                if (onCancel) {
                    return <CancelLibrary onClick={onCancel} />
                }
                return <Cancel center link={previousPath} />
            }
        }
    }

    const formFields =
        activeTab === 'tip' && subCategory ? subCategory.fields : (activeTab === 'tour' && tourFormFields) || hotelFormFields
    const formIcon = activeTab === 'tip' ? 'heart' : (activeTab === 'tour' && 'tour') || 'hotel'

    return (
        <>
            <WrapperPadding>
                <Panel
                    backgroundImageUrl={tip && tip.getImage && tip.getImage()}
                    icon={formIcon}
                    onArquive={tipId && del}
                    requiredImage={!hasImage}
                    uploadImage={updateImage}
                >
                    <If condition={showTabs}>
                        <Tabs activeTab={activeTab} updateActiveTab={updateActiveTab} />
                    </If>
                </Panel>
            </WrapperPadding>
            <If condition={activeTab === 'tip'}>
                <Select categories={categories} onChange={handleSelectCategory} value={tip && tip.subCategory} />
            </If>
            <If condition={isNotEmpty(category) || activeTab !== 'tip'}>
                <FormBuilder
                    disabled={_formDisabled}
                    fields={formFields}
                    getSuggestions={requestSearchCityOurApiAction}
                    getSuggestions2={requestSearchCityGoogleApiAction}
                    initialValues={tip ? tip.toForm() : { boarding: { date: scriptDate } }}
                    onSubmit={callSaveData}
                    ref={process.env.__TEST__ ? null : form}
                    suggestions={suggestions}
                />
                {hasImage && (
                    <FormButtonContainer>
                        <If
                            condition={_formDisabled}
                            el={
                                <Button
                                    color={colors.purple}
                                    disabled={false}
                                    onClick={onSaveClick}
                                    text={i18n.t('buttons.saveService')}
                                />
                            }
                        >
                            <Button
                                color={colors.purple}
                                disabled={false}
                                onClick={() => _updateFormDisabled(false)}
                                text={i18n.t('buttons.editService')}
                            />
                        </If>
                    </FormButtonContainer>
                )}
            </If>
            {renderCancel()}
        </>
    )
}
/* eslint-enable complexity */
TipForm.defaultProps = {
    formDisabled: true,
    onCancel: null,
    redirect: true,
    showCancel: true,
    showTabs: false,
    suggestions: [],
    tip: {}
}

TipForm.propTypes = {
    formDisabled: PropTypes.bool,
    match: PropTypes.object.isRequired,
    onCancel: PropTypes.func,
    redirect: PropTypes.bool,
    requestDelAction: PropTypes.func.isRequired,
    requestGetCategoryAction: PropTypes.func.isRequired,
    requestGetSingleAction: PropTypes.func.isRequired,
    requestSaveDataAction: PropTypes.func.isRequired,
    requestSearchCityGoogleApiAction: PropTypes.func.isRequired,
    requestSearchCityOurApiAction: PropTypes.func.isRequired,
    showCancel: PropTypes.bool,
    showTabs: PropTypes.bool,
    suggestions: PropTypes.array,
    tip: PropTypes.objectOf(TipServiceModel)
}

export default withRouter(TipForm)
