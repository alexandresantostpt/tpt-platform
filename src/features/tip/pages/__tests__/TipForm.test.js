import React from 'react'
import renderer from 'react-test-renderer'

import TipServiceModel from '../../models/TipServiceModel'

import TipForm from '../TipForm'

describe('Tests for TipForm component', () => {
    const tip = new TipServiceModel({
        __v: 0,
        _id: '5c52ebcd5fc7c9b1433ab704',
        address: 'Test',
        category: 'Category',
        describe: 'Describe',
        observations: 'Test',
        phone: '01934568945',
        script: '5c52ebcd5fc7c9b1433ab704',
        scriptDate: '2019-03-12T20:00:00.458Z',
        site: 'site',
        subCategory: 'subCategory',
        title: 'Title',
        type: 'Type',
        workDays: 'From monday to thursday',
        workEnd: '2019-03-12T20:00:00.458Z',
        workStart: '2019-03-12T08:00:00.458Z'
    })

    it('Should render withou crash', () => {
        const match = {
            params: {
                id: '5c52ebcd5fc7c949223ab705'
            }
        }
        const tree = renderer
            .create(
                <TipForm
                    match={match}
                    requestDelAction={jest.fn()}
                    requestGetSingleAction={jest.fn()}
                    requestSaveDataAction={jest.fn()}
                    tip={tip}
                />
            )
            .toJSON()
        expect(tree).toMatchSnapshot()
    })
})
