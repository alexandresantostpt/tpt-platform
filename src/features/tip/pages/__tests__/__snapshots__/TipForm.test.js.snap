// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for TipForm component Should render withou crash 1`] = `
Array [
  <div
    className="WrapperPadding-ovjbyx-0 lcnoja"
  >
    <div
      className="Panel__StyledPanel-sc-1qzkpb6-0 cGUlCn"
    >
      <span
        className="Panel__StyledIcon-sc-1qzkpb6-2 iavvSk"
      >
        <span
          className="icon icon-heart Icon__StyledIcon-crr1cg-0 dfhcZR"
          disabled={false}
          fontSize="3rem"
          onClick={null}
        />
      </span>
      <div
        className="UploadButton__StyledContainer-sc-1p8zoc-0 fpZFIG"
      >
        <label
          className="UploadButton__StyledUploadButton-sc-1p8zoc-3 ivhrYW"
          htmlFor="inputFile"
        >
          +
        </label>
        <input
          accept="image/png, image/jpeg"
          className="UploadButton__InputFile-sc-1p8zoc-4 eLaooA"
          id="inputFile"
          onChange={[Function]}
          type="file"
        />
      </div>
    </div>
  </div>,
  <div
    className="Select__StyledSelect-sc-1jg94wo-0 hwuoao"
    onClick={[Function]}
  >
    <div
      className="SelectContainer__StyledSelectContainer-b4xrfw-0 dbBaBH"
    >
      <span
        className="SelectText-sc-9v8jof-0 kzQEMd"
        onClick={[Function]}
      >
        Selecione uma categoria...
        <span
          className="icon icon-arrow-down Icon__StyledIcon-crr1cg-0 cmjVJQ"
          disabled={false}
          fontSize="1.75rem"
          onClick={null}
        />
      </span>
    </div>
    <ul
      className="SelectList__StyledSelectList-sc-13lmtnk-0 kdpUHJ"
    >
      <span>
        <li
          className="GroupItem__StyledGroupItem-sc-1p2pi5i-0 gXBQCx"
        >
          Eventos
        </li>
      </span>
      <span>
        <li
          className="GroupItem__StyledGroupItem-sc-1p2pi5i-0 gXBQCx"
        >
          Gastronomia
        </li>
      </span>
      <span>
        <li
          className="GroupItem__StyledGroupItem-sc-1p2pi5i-0 gXBQCx"
        >
          Vida noturna
        </li>
      </span>
      <span>
        <li
          className="GroupItem__StyledGroupItem-sc-1p2pi5i-0 gXBQCx"
        >
          Compras
        </li>
      </span>
      <span>
        <li
          className="GroupItem__StyledGroupItem-sc-1p2pi5i-0 gXBQCx"
        >
          Passeio
        </li>
      </span>
    </ul>
  </div>,
]
`;
